include/oneapi/dpl/internal/random_impl/philox_engine.h

```diff
@@ -0,0 +1,483 @@
+        return __ret;
+    }
+
+    /* Shift the counter only forward relative to its current position */
+    void
+    discard(unsigned long long __z)
+    {
+        discard_internal(__z);
+    }
+
+    /* Equality operators */
+    friend bool
+    operator==(const philox_engine& __x, const philox_engine& __y)
+    {
+        if (!::std::equal(__x.state_.X.begin(), __x.state_.X.end(), __y.state_.X.begin()) ||
+            !::std::equal(__x.state_.K.begin(), __x.state_.K.end(), __y.state_.K.begin()) ||
+            !::std::equal(__x.state_.Y.begin(), __x.state_.Y.end(), __y.state_.Y.begin()) ||
+            __x.state_.idx != __y.state_.idx)
+        {
+            return false;
+        }
+        return true;
+    }
+    friend bool
+    operator!=(const philox_engine& __x, const philox_engine& __y)
+    {
+        return !(__x == __y);
+    }
+
+    /* Inserters and extractors */
+    template <class _CharT, class _Traits, typename __UIntType, ::std::size_t __w, ::std::size_t __n, ::std::size_t __r,
+              __UIntType... __consts>
+    friend ::std::basic_ostream<_CharT, _Traits>&
+    operator<<(::std::basic_ostream<_CharT, _Traits>&, const philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+    template <typename __UIntType, ::std::size_t __w, ::std::size_t __n, ::std::size_t __r, __UIntType... __consts>
+    friend const sycl::stream&
+    operator<<(const sycl::stream&, const philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+    template <class _CharT, class _Traits, typename __UIntType, ::std::size_t __w, ::std::size_t __n, ::std::size_t __r,
+              __UIntType... __consts>
+    friend ::std::basic_istream<_CharT, _Traits>&
+    operator>>(::std::basic_istream<_CharT, _Traits>&, philox_engine<__UIntType, __w, __n, __r, __consts...>&);
+
+  private:
+    /* Internal generator state */
+    struct state
+    {
+        ::std::array<scalar_type, word_count> X;     // counters
+        ::std::array<scalar_type, word_count / 2> K; // keys
+        ::std::array<scalar_type, word_count> Y;     // results
+        scalar_type idx;                             // index
+    } state_;
+
+    /* Processing mask */
+    static constexpr auto in_mask = internal::experimental::word_mask<scalar_type, word_size>;
+
+    void
+    seed_internal(::std::initializer_list<scalar_type> __seed)
+    {
+        auto __start = __seed.begin();
+        auto __end = __seed.end();
+        // all counters are set to zero
+        for (::std::size_t __i = 0; __i < word_count; __i++)
+        {
+            state_.X[__i] = 0;
+        }
+        // keys are set as seed
+        for (::std::size_t __i = 0; __i < (word_count / 2); __i++)
+        {
+            state_.K[__i] = (__start == __end) ? 0 : (*__start++) & in_mask;
+        }
+        // results are set to zero
+        for (::std::size_t __i = 0; __i < word_count; __i++)
+        {
+            state_.Y[__i] = 0;
+        }
+
+        state_.idx = word_count;
```

Comment:
we are not aligned with paper here (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p2075r6.pdf): according to the paper we need to set idx as word_count -1 and when we do generate, first increment counter, then check idx == word_count:
![image](https://github.com/user-attachments/assets/ed905761-94d8-475a-bfdb-a32531593142)
Ideally we need to align but it will cause logic shift in generate and discard

Reply:
The logic of idx iteration is changed as well as the generation methods and discard method.